Initial implementation of GraphBLAS dialect

paul-tqh-nguyen opened the pull request:
This PR contains scaffolding for the GraphBLAS out-of-tree MLIR dialect.

`graphblas-opt` and `graphblas-translate` can be built using the `build.py` script.

Commit:
Initial commit of GraphBLAS dialect

This commit contains scaffolding for the GraphBLAS out-of-tree MLIR dialect.

seibert commented on dev-environment.yml:
```diff
@@ -30,3 +30,6 @@ dependencies:
   - panel
   - bokeh<2.3  # temp restriction until panel/bokeh bugs are worked out
+  - cmake >= 3.13.4
+  - conda-forge::ninja
```
Is there a reason we need ninja from conda-forge?  It looks like the same versions are available in defaults as well.

paul-tqh-nguyen replied:
https://github.com/metagraph-dev/mlir-graphblas/pull/38/commits/053d56814992458df38906aa914637ea579ec2f6 removes the use of `conda-forge` for `ninja`.

seibert commented on dev-environment.yml:
```diff
@@ -30,3 +30,6 @@ dependencies:
   - bokeh<2.3  # temp restriction until panel/bokeh bugs are worked out
+  - cmake >= 3.13.4
+  - conda-forge::ninja
+  - conda-forge::lit
```
Same question here, but conda-forge has newer versions of `lit` than defaults.  Do we need one of those newer versions?  (In which case we should set a version constraint.)

paul-tqh-nguyen replied:
https://github.com/metagraph-dev/mlir-graphblas/pull/38/commits/053d56814992458df38906aa914637ea579ec2f6 removes the use of `conda-forge` for `lit`.

seibert commented on mlir_graphblas/src/CMakeLists.txt:
```diff
@@ -0,0 +1,35 @@
+add_subdirectory(lib)
+add_subdirectory(test)
+add_subdirectory(graphblas-opt)
+add_subdirectory(graphblas-translate)
```
Do we need our own `translate` command?  I don't think we'll be customizing that.

paul-tqh-nguyen replied:
https://github.com/metagraph-dev/mlir-graphblas/pull/38/commits/aad0144c102c86b13b3d8861b7206b33dba6ddb4 removes `graphblas-translate`.

seibert commented on mlir_graphblas/src/build.py:
```diff
@@ -0,0 +1,88 @@
+        os.makedirs(build_dir)
+
+    env_lib_path = os.path.join(sys.exec_prefix, "lib")
+    LD_LIBRARY_PATH = os.environ.get("LD_LIBRARY_PATH", "")
+    if env_lib_path not in LD_LIBRARY_PATH.split(":"):
+        LD_LIBRARY_PATH = LD_LIBRARY_PATH + ":" + env_lib_path
+
+    (PYTHONPATH,) = site.getsitepackages()
+
+    command, stdout_string, stderr_string = run_shell_commands(
+        build_dir,
+        "cmake -G Ninja .. -DMLIR_DIR=$PREFIX/lib/cmake/mlir -DLLVM_EXTERNAL_LIT=$BUILD_DIR/bin/llvm-lit",
+        "cmake --build . --target check-graphblas",
+        PREFIX=sys.exec_prefix,
+        BUILD_DIR=sys.exec_prefix,
+        LD_LIBRARY_PATH=LD_LIBRARY_PATH,
+        PYTHONPATH=PYTHONPATH,
```
Note that `LD_LIBRARY_PATH` doesn't work on macOS, where it should be `DYLD_LIBRARY_PATH`.  Does this work on the Mac?

paul-tqh-nguyen replied:
According to [this article](https://medium.com/macos-is-not-linux-and-other-nix-reflections/d-o-y-ou-ld-library-path-you-6ab0a6135a33), `LD_LIBRARY_PATH` should work. 


`build.py` works with `LD_LIBRARY_PATH`.
If I instead use `DYLD_LIBRARY_PATH`, it doesn’t work. 
However, using `DYLD_FALLBACK_LIBRARY_PATH` works.

I'm okay with implementing this whatever way seems best. Using `LD_LIBRARY_PATH` doesn't seem to cause any problems.

seibert replied:
if this works, I'm happy to leave it

seibert commented on mlir_graphblas/src/graphblas-opt/graphblas-opt.cpp:
```diff
@@ -0,0 +1,36 @@
+  // Add the following to include *all* MLIR Core dialects, or selectively
+  // include what you need like above. You only need to register dialects that
+  // will be *parsed* by the tool, not the one generated
+  // registerAllDialects(registry);
```
Should we include all dialects?  I know we can chain this tool with the standard `mlir-opt` to get the rest of the dialects, but are there any downsides to an all-in-one tool?

paul-tqh-nguyen replied:
I don't see any downsides to creating an all-in-one tool.

This comment was left here as a result of using the [`standalone` MLIR template](https://github.com/llvm/llvm-project/blob/e7021232e66f4a8e21f2bcd77d9841d1fb414245/mlir/examples/standalone/standalone-opt/standalone-opt.cpp#L31). 

Perhaps it might be a good idea to decide on this when we write our first set of passes.

seibert reviewed (changes requested):
Looks good!  The main high level question is whether we need `graphblas-translate` at all, since translations are for going from MLIR to another IR (like LLVM), and we should be able to use the standard translate for everything we do.

Also, it would be good to add instructions to the readme about how to test `graphblas-opt`.

Commit:
Remove graphblas-translate

This commit removes graphblas-translate and updates the dev-environment.yml file.

Commit:
Add graphblas-opt instructions to README

This adds instructions on how to compile and test graphblas-opt to the README as well as how to perform a clean build.

paul-tqh-nguyen commented:
@seibert , I've implemented the changes you've suggested. 

One thing to note is that this PR relies on an MLIR conda package built from [this recipe](https://github.com/metagraph-dev/recipes/commit/961baa1a33acac89515ffad0526900359fbfc86d), which is not [the latest package on the metagraph channel](https://anaconda.org/metagraph/mlir-graphblas).

I'm not exactly sure what we want to do with `LD_LIBRARY_PATH`, but I'm happy to have the build script use whatever environment variable seems to be the best practice.

https://github.com/metagraph-dev/mlir-graphblas/pull/38/commits/055f9c15db1c676cfed80c88a94437e6b5ff2f9e adds instructions to the README on how to build `graphblas-opt` and how to run tests. Running tests is necessary to building `graphblas-opt` (the command to run the tests is needed to build `mlir_graphblas/src/build/bin/graphblas-opt` ; I haven't figured out how to build this file without using [the command to run the tests](https://github.com/metagraph-dev/mlir-graphblas/pull/38/commits/055f9c15db1c676cfed80c88a94437e6b5ff2f9e#diff-da14321ddb1a3c1b448b244ad8d943bb72c8d610e7ed44effb912129f30b7334R81)) and building `graphblas-opt` is necessary for running the tests (for obvious reasons). Thus, just building `graphblas-opt` will run the tests. I've also added command-line args to build from scratch. 

I think it's ready to merge unless there's something I should do with `LD_LIBRARY_PATH`.

seibert commented:
Looks good!  I will update the mlir packages on the metagraph channel.
